dialer.go

```diff
@@ -81,137 +83,161 @@ func NewDialer(opts *DialerOpts) Dialer {
 	log.Debugf("Initializing Dialer with   windowSize: %v   maxPadding: %v   maxLiveConns: %v  maxStreamsPerConn: %v   pingInterval: %v   cipher: %v",
 		opts.WindowSize,
 		opts.MaxPadding,
-		opts.MaxLiveConns,
+		opts.LiveConns,
 		opts.MaxStreamsPerConn,
 		opts.PingInterval,
 		opts.Cipher)
-	liveSessions := make(chan sessionIntf, opts.MaxLiveConns)
-	liveSessions <- nullSession{}
-	return &dialer{
+	var lc ClientLifecycleListener
+	if opts.Lifecycle != nil {
+		lc = opts.Lifecycle
+	} else {
+		lc = NoopClientLifecycleListener()
+	}
+	d := &dialer{
 		windowSize:            opts.WindowSize,
 		maxPadding:            opts.MaxPadding,
 		maxStreamsPerConn:     opts.MaxStreamsPerConn,
-		maxLiveConns:          opts.MaxLiveConns,
 		idleInterval:          opts.IdleInterval,
 		pingInterval:          opts.PingInterval,
 		redialSessionInterval: opts.RedialSessionInterval,
-		pool:            opts.Pool,
-		cipherCode:      opts.Cipher,
-		serverPublicKey: opts.ServerPublicKey,
-		liveSessions:    liveSessions,
-		numLive:         1, // the nullSession
-		emaRTT:          ema.NewDuration(0, 0.5),
+		pool:                  opts.Pool,
+		cipherCode:            opts.Cipher,
+		serverPublicKey:       opts.ServerPublicKey,
+		liveSessions:          make(chan sessionIntf, opts.LiveConns),
+		emaRTT:                ema.NewDuration(0, 0.5),
+		dial:                  opts.Dial,
+		pendingSessions:       make(chan *pendingSession, opts.LiveConns),
+		lifecyle:              lc,
+	}
+	d.lifecyle.OnStart()
+	for i := 0; i < opts.LiveConns-1; i++ {
+		d.pendingSessions <- newPendingSession()
 	}
+
+	// We create another background session with a shorter dial timeout to ensure liveness in the case of network
+	// disruptions.
+	d.pendingSessions <- &pendingSession{
+		name:         "liveness",
+		dialTimeout:  5 * time.Second,
+		sleepOnError: 0 * time.Second,
```

Comment:
Oh I just realized when seeing logs - if dialing lampshade gets RST or errors like no route to host/network, Lantern would busily dial new session over and over again without sleeping.